server/src/main/java/io/unitycatalog/server/service/ModelService.java

```diff
@@ -3,102 +3,260 @@
     RegisteredModelInfo updateRegisteredModelResponse =
         MODEL_REPOSITORY.updateRegisteredModel(updateRegisteredModel);
     return HttpResponse.ofJson(updateRegisteredModelResponse);
   }
 
   @Delete("/{full_name}")
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          (#authorize(#principal, #registered_model, OWNER) && #authorize(#principal, #schema, USE_SCHEMA) && #authorize(#principal, #catalog, USE_CATALOG))
+          """)
+  @AuthorizeKey(METASTORE)
   public HttpResponse deleteRegisteredModel(
-      @Param("full_name") String fullName, @Param("force") Optional<Boolean> force) {
+      @Param("full_name") @AuthorizeKey(REGISTERED_MODEL) String fullName, @Param("force") Optional<Boolean> force) {
+    RegisteredModelInfo registeredModelInfo = MODEL_REPOSITORY.getRegisteredModel(fullName);
     MODEL_REPOSITORY.deleteRegisteredModel(fullName, force.orElse(false));
+    removeAuthorizations(registeredModelInfo);
     return HttpResponse.of(HttpStatus.OK);
   }
 
   @Post("/versions")
-  public HttpResponse createModelVersion(CreateModelVersion createModelVersion) {
+  @AuthorizeExpression("""
+          #authorize(#principal, #catalog, USE_CATALOG) && #authorizeAll(#principal, #schema, USE_SCHEMA, CREATE_MODEL)
+          """)
+  public HttpResponse createModelVersion(@AuthorizeKeys({
+          @AuthorizeKey(value = CATALOG, key = "catalog_name"),
+          @AuthorizeKey(value = SCHEMA, key = "schema_name"),
+          @AuthorizeKey(value = REGISTERED_MODEL, key = "model_name")})
+          CreateModelVersion createModelVersion) {
     assert createModelVersion != null;
     ModelVersionInfo createModelVersionResponse =
         MODEL_REPOSITORY.createModelVersion(createModelVersion);
     return HttpResponse.ofJson(createModelVersionResponse);
   }
 
   @Get("/{full_name}/versions")
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          #authorize(#principal, #registered_model, OWNER) ||
+          (#authorize(#principal, #registered_model, EXECUTE) && #authorize(#principal, #schema, USE_SCHEMA) && #authorize(#principal, #catalog, USE_CATALOG))
+          """)
+  @AuthorizeKey(METASTORE)
   public HttpResponse listModelVersions(
-      @Param("full_name") String fullName,
+      @Param("full_name") @AuthorizeKey(REGISTERED_MODEL) String fullName,
       @Param("max_results") Optional<Integer> maxResults,
       @Param("page_token") Optional<String> pageToken) {
     return HttpResponse.ofJson(MODEL_REPOSITORY.listModelVersions(fullName, maxResults, pageToken));
   }
 
   @Get("/{full_name}/versions/{version}")
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          #authorize(#principal, #registered_model, OWNER) ||
+          (#authorize(#principal, #registered_model, EXECUTE) && #authorize(#principal, #schema, USE_SCHEMA) && #authorize(#principal, #catalog, USE_CATALOG))
+          """)
```

Comment:
Metastore, catalog, and schema owners should be able to get a model version.

Reply:
I added explicit checks for catalog and schema OWNER